sys/arm64/arm64/pmap.c

```diff
@@ -709,26 +709,25 @@ pmap_pte_prot(pmap_t pmap, vm_prot_t prot)
 			val |= ATTR_S1_XN;
 		if ((prot & VM_PROT_WRITE) == 0)
 			val |= ATTR_S1_AP(ATTR_S1_AP_RO);
-#if __has_feature(capabilities)
-		if ((prot & VM_PROT_READ_CAP) != 0)
-			val |= ATTR_LC_ENABLED;
-		if ((prot & VM_PROT_WRITE_CAP) != 0)
-			val |= ATTR_SC;
-#endif
 	} else {
 		if ((prot & VM_PROT_WRITE) != 0)
 			val |= ATTR_S2_S2AP(ATTR_S2_S2AP_WRITE);
 		if ((prot & VM_PROT_READ) != 0)
 			val |= ATTR_S2_S2AP(ATTR_S2_S2AP_READ);
 		if ((prot & VM_PROT_EXECUTE) == 0)
 			val |= ATTR_S2_XN(ATTR_S2_XN_ALL);
+	}
+
 #if __has_feature(capabilities)
-		if ((prot & VM_PROT_READ_CAP) != 0)
-			val |= ATTR_LC_ENABLED;
-		if ((prot & VM_PROT_WRITE_CAP) != 0)
-			val |= ATTR_SC;
+	if ((prot & VM_PROT_READ_CAP) != 0)
+		val |= ATTR_LC_ENABLED;
+
+	VM_PAGE_ASSERT_PGA_CAPMETA_PMAP_ENTER(m, prot);
+	if ((prot & VM_PROT_WRITE_CAP) != 0)
+		val |= ATTR_CDBM;
+	if ((flags & VM_PROT_WRITE_CAP) != 0)
+		val |= ATTR_SC;
```

Comment:
The SC bit was included to match the behavior of the riscv pmap_enter: if the flags argument to that function has VM_PROT_WRITE_CAP, then the pte has the capdirty bit set.

As for CDBM, the corresponding code that sets DBM is further down in pmap_enter: line 3925 in this diff. I thought it would make more sense to keep the capability related stuff together (though I've just lifted the assert out of pmap_pte_prot and into pmap_enter in line with your above comment).